src/runtime_src/core/edge/user/aie/aie.cpp

```diff
@@ -392,55 +477,183 @@ start_profiling(int option, const std::string& port1_name, const std::string& po
   return handle;
 }
 
-uint64_t
+int
 Aie::
-read_profiling(int phdl)
+start_profiling_start_bytes(const std::string& port_name, uint32_t value)
 {
-  uint64_t value = 0;
+  int handle = -1;
 
-  std::vector<Resources::AcquiredResource>& acquiredResourcesForThisHandle = eventRecords[phdl].acquiredResources;
+  XAie_LocType shim_tile;
+  XAie_StrmPortIntf mode;
+  uint8_t stream_id;
 
-  Resources::AcquiredResource& acquiredResource = acquiredResourcesForThisHandle[0];
-  XAie_ModuleType XAieModuleType = AIEResourceModuletoXAieModuleTypeMap[acquiredResource.module];
+  get_profiling_config(port_name, shim_tile, mode, stream_id);
 
-  if (acquiredResource.resource == Resources::performance_counter)
-    XAie_PerfCounterGet(devInst, acquiredResource.loc, XAieModuleType, acquiredResource.id, (u32*)(&value));
-  else
-    throw xrt_core::error(-EAGAIN, "Can't read profiling: The acquired resources order does not match the profiling option.");
+  int handleId = eventRecords.size();
+  int eventPortId = Resources::AIE::getShimTile(shim_tile.Col)->plModule.requestStreamEventPort(handleId);
+  int counterId0 = Resources::AIE::getShimTile(shim_tile.Col)->plModule.requestPerformanceCounter(handleId);
+  int counterId1 = Resources::AIE::getShimTile(shim_tile.Col)->plModule.requestPerformanceCounter(handleId);
 
-  return value;
+  if (counterId0 >= 0 && counterId1 >= 0 && eventPortId >= 0) {
+    XAie_EventSelectStrmPort(devInst, shim_tile, (uint8_t)eventPortId, mode, SOUTH, stream_id);
+    XAie_PerfCounterEventValueSet(devInst, shim_tile, XAIE_PL_MOD, (uint8_t)counterId1, value / 4);
+    XAie_PerfCounterControlSet(devInst, shim_tile, XAIE_PL_MOD, (uint8_t)counterId0, XAIETILE_EVENT_SHIM_PORT_RUNNING[eventPortId], XAIE_EVENT_PERF_CNT_1_PL);
+    XAie_PerfCounterControlSet(devInst, shim_tile, XAIE_PL_MOD, (uint8_t)counterId1, XAIETILE_EVENT_SHIM_PORT_RUNNING[eventPortId], XAIETILE_EVENT_SHIM_PORT_RUNNING[eventPortId]);
+    eventRecords.push_back( { IO_STREAM_START_TO_BYTES_TRANSFERRED_CYCLES,
+                { { shim_tile, Resources::pl_module, Resources::performance_counter, (size_t)counterId0 },
+                { shim_tile, Resources::pl_module, Resources::performance_counter, (size_t)counterId1 },
+                { shim_tile, Resources::pl_module, Resources::stream_switch_event_port, (size_t)eventPortId } } } );
```

Comment:
Return type of various APIs and how these are used don't seem to agree very well.  Why are there so many casts?  Also C-style casting is not good.  Please use static_cast, more type yes, but more incentive to fix and easier to locate bad code.